### PR TITLE
Add `LinkModel` class with CRU operations

### PR DESCRIPTION
**Pull request by thehabbos007**
## 🌟 What is the purpose of this PR?

This PR adds a new link model class that allows for CRU  operations on Links in the Graph API.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1202787635216884/1202812736735648/f) _(internal)_
- Fixes [Asana task](https://app.asana.com/0/1202805690238892/1202883599104668/f) _(internal)_ transitively

## 🔍 What does this change?

- Adds a `link.model.ts` file allowing for the Graph API Link operations. 
- Adds helper methods to `entity.model.ts` that delegates to these new operations
- Tests these methods against the DB
- Changes `Links` to `OutgoingLinks` in the Graph API, making it a tuple struct instead of relying on flatten
    - Renames `Outgoing` enum to `OutgoingLinkTarget`
- Changes `Entity` into a tuple struct instead of relying on flatten (due to validations introduced on base URIs https://github.com/blockprotocol/blockprotocol/commit/c0a430dfd8d5108272c41c0894a3d1ce0b6e1bd8)

## 🐾 Next steps

- Migration organisations [Asana task](https://app.asana.com/0/0/1202606867065366/f) (_internal_)
- Allowing for 1:N links in the datastore. [Asana task](https://app.asana.com/0/1201095311341924/1202681321906843/f) (_internal_)

## 🛡 What tests cover this?

- Integration tests at `packages/hash/integration/src/tests/model/knowledge/link.model.test.ts`

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Recreate and migrate DB and run backend
     -  From the root run `yarn external-services up --build` (the `--build` arg here is important as the backend has been updated)
1.  In another terminal, from the `packages/hash/integration` path run `yarn jest "knowledge/link"`
1.  Confirm all tests pass

## 📹 Demo

image

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs [hunk 1]
```diff
@@ -67,8 +68,8 @@ pub enum Outgoing {
 
 /// A collection of links that originate from the same source entity.
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Component)]
+#[serde(rename_all = "camelCase")]
 pub struct Links {
-    #[serde(flatten)]
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs:71]: Why did we get rid of this

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs:71]: Was this to do with the OpenAPI library not being able to deal with flatten or something?

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
+import { GraphApi } from "@hashintel/hash-graph-client";
+
+import { EntityModel, LinkModel, LinkTypeModel } from "../index";
+
+export type LinkModelConstructorParams = {
+  accountId: string;
+  sourceEntity: EntityModel;
```
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/link.model.ts:7]: sourceEntityModel*? (and target)

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  /**
+   * Create a link between a source and a target entity using a specific link
+   * type.
+   *
+   * @param params.accountId the accountId of the account creating the link
+   * @param params.sourceEntity the source entity of the link
+   * @param params.linkTypeModel the Link Type of the link
+   * @param params.targetEntity the target entity of the link
+   */
+  static async create(
```
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/link.model.ts:51]: I think this model is the most confusing to me, why is it possible to instantiate a model without the link being created, and if it is already created then why do we have a `create` method? I think my confusion is just stemming from me not understanding the point of the model classes properly in the first place

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  ): Promise<LinkModel[]> {
+    const { data: entityLinks } = await graphApi.getEntityLinks(
+      sourceEntity.entityId,
+    );
+    return Promise.all(
+      Object.entries(entityLinks.outgoing)
+        /**
+         * When target entity id is a list the link has many targets.
+         * This is not supported yet
+         * @todo support 1:N links
```
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/link.model.ts:106]: I don't understand this, why do we keep referring to this as a single link to multiple things? Is that how it was modelled prior to this rework?

I see this as just multiple links, where they all happen to share the same source. 

I don't see _**any**_ value of thinking about _a_ link having multiple targets. It seems like it would greatly confuse versioning, representations, and indeed this model class wouldn't make sense because it has `targetEntityModel` which is a singular thing.

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+         * */
+        .filter(([_, targetEntityId]) => typeof targetEntityId === "string")
```
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/link.model.ts:108]: Why is this necessary

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        .map(
+          async ([linkTypeUri, targetEntityId]) =>
+            new LinkModel({
+              accountId: sourceEntity.accountId,
+              linkTypeModel: await LinkTypeModel.get(graphApi, {
+                versionedUri: linkTypeUri,
+              }),
+              sourceEntity,
+              targetEntity: await EntityModel.getLatest(graphApi, {
+                /** @todo figure out what account ID we use here */
```
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/link.model.ts:118]: When are we going to do these todos? I'm slightly concerned that we're throwing a lot of todos around the codebase without thinking too careful about their discoverability. We should probably start embedding their relevant asana links directly in the todo just so that when we pick up the asana task we can `ctrl+f` for all occurrences or something.

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+                accountId: sourceEntity.accountId,
+                entityId: targetEntityId,
+              }),
+            }),
+        ),
+    );
+  }
+
+  /**
+   * Get the outgoing link of a source entity given a link type.
```
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/link.model.ts:128]: I think this is similar to the above but I don't see how this can be guaranteed to return just one thing. What if there are multiple `Friend` links going from `Bob`

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+      linkTypeModel,
+    }: { sourceEntity: EntityModel; linkTypeModel: LinkTypeModel },
+  ): Promise<LinkModel | null> {
+    /** @todo use structural querying for this client-side fetch */
+    const links = (
+      await LinkModel.getAllOutgoing(graphApi, { sourceEntity })
+    ).filter(
+      (link) => link.linkTypeModel.schema.$id === linkTypeModel.schema.$id,
+    );
+    /** @todo the may return an array of links whe nwe support 1:N links. */
```
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/link.model.ts:147]: Typo but also see above

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+   * @param params.linkTypeModel the Link Type of the link
+   */
+  static async getOutgoing(
+    graphApi: GraphApi,
+    {
+      sourceEntity,
+      linkTypeModel,
+    }: { sourceEntity: EntityModel; linkTypeModel: LinkTypeModel },
+  ): Promise<LinkModel | null> {
+    /** @todo use structural querying for this client-side fetch */
```
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/link.model.ts:141]: Similar comment about attaching this TODO to something

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+      await LinkModel.getAllOutgoing(graphApi, { sourceEntity })
+    ).filter(
+      (link) => link.linkTypeModel.schema.$id === linkTypeModel.schema.$id,
+    );
+    /** @todo the may return an array of links whe nwe support 1:N links. */
+    return links[0] ? links[0] : null;
+  }
+
+  /**
+   * Make a link inactive.
```
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/link.model.ts:152]: For my understanding, what is an inactive link?

File: packages/hash/integration/src/tests/model/knowledge/link.model.test.ts [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+  let targetEntityFriend: EntityModel;
+  let targetEntityAcquaintance: EntityModel;
+
+  beforeAll(async () => {
+    testType = await EntityTypeModel.create(graphApi, {
+      accountId,
+      schema: {
+        $id: entityType$id,
+        kind: "entityType",
+        title: "Text",
```
Comment (Alfred-Mountfield) [inline packages/hash/integration/src/tests/model/knowledge/link.model.test.ts:46]: I know this is just a test and we've been pretty poor about adding meaningful types but I think calling an _entity_ type `Text is even more confusing than is necessary at this point 😄 can we call it something else?

File: packages/hash/integration/src/tests/model/knowledge/link.model.test.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        type: "object",
+        properties: {},
+      },
+    });
+
+    const results = await Promise.all([
+      LinkTypeModel.create(graphApi, {
+        accountId,
+        schema: {
+          $id: "https://link~example.com/link-types/friends-with/v/1",
```
Comment (Alfred-Mountfield) [inline packages/hash/integration/src/tests/model/knowledge/link.model.test.ts:56]: Is this really a valid URL? I find it quite surprising that you can use a tilde.

File: packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs (hunk 1 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs:71]: Yes, exactly

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 2 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts:7]: Yes, that should have been it :)

File: packages/hash/integration/src/tests/model/knowledge/link.model.test.ts (hunk 12 above)
Comment (thehabbos007) [reply packages/hash/integration/src/tests/model/knowledge/link.model.test.ts:56]: image
Seems valid, yes. I think it would not be accessible, though. It would expand to home directory or something

File: packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs:71]: @TimDiekmann any concerns about this?

Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs:71]: As long as the tests are passing I don't mind changing the `serde` attributes 🚀

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs:71]: Well we've updated the tests here so that might not be the best criteria 😁 

Is there a reason we have the `outgoing` field? Could we perhaps change it to an untagged new-type to avoid the flatten?

`OutgoingLinks(HashMap<...`?

Or did we intend to add more fields

Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs:71]: If we intend to have `incoming` as well, it would make sense for this struct to be as it is. I don't remember how utoipa handles newtypes that are not just strings, but it might have difficulties when it's a `HashMap`

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 10 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts:152]: It's more or less a soft-deleted link

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 5 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts:108]: The endpoint returns `{ [string]: string | string[] }`, so this is to only deal with the singular links

File: packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs:71]: Would we put `incoming` in here? 
And if there are other solutions that we can explore I think we should. I'd prefer to avoid polluting these structs if we can.

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 4 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts:106]: I refer to it as 1:N where one source entity links to multiple targets under the same link type. Not a single link to multiple things. I would like a better name for this, but I've settled with "1:N links" for now.

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 10 above)
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/knowledge/link.model.ts:152]: Interesting, I think we should revisit this naming (not in this PR)

File: packages/hash/integration/src/tests/model/knowledge/link.model.test.ts (hunk 12 above)
Comment (Alfred-Mountfield) [reply packages/hash/integration/src/tests/model/knowledge/link.model.test.ts:56]: Strange. I think we're going to have to revisit all of our test URIs anyway soon as we're going to be putting even more validation (ensuring it's a domain we can serve from, etc.) so no worries right now.

File: packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs (hunk 1 above)
Comment (thehabbos007) [reply packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs:71]: I suppose we wouldn't put it in here, as it might be a structural query result to get all incoming links. Do you want me to see what happens if we use a newtype instead?

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 3 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts:51]: This method is static, so it wouldn't be relevant to instances. I do agree that we shouldn't allow instantiating a model without the link being created, perhaps we should aim for making the constructor `private` or something similar? (I think that might mess with some of the inheritance stuff being done in the model class, but I can make it `private` for now)

File: packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/knowledge/link/mod.rs:71]: I think if we are not planning on adding `incoming` here (which would probably be odd), avoiding the `.outgoing` would be good I guess.

**Comment by thehabbos007** [comment]
A little bit of a detour since the review @Alfred-Mountfield. When trying to run tests, everything failed because of changes to the graph API in the meanwhile. Since the new type system packages do better validation of entity property URIs, all tests were failing because of a lack of a trailing slash on base URIs (https://github.com/blockprotocol/blockprotocol/commit/c0a430dfd8d5108272c41c0894a3d1ce0b6e1bd8). 
I had to fix the transitive errors by this drive-by: 1581a988d89ce4ce0b366a64a5dc2df1fe8b7e23
and then fix existing tests using Entity insertion/update/fetch: 3dfff079f8414285ba4f6e82fbc6c5d69db665eb

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 6 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts]: Added link to task in ce907a39c4449b548374a9542a9533c9a93bdf37

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 9 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts]: Addressed in 6e2460441de91db3197537fdc273e40689f83cee

File: packages/hash/integration/src/tests/model/knowledge/link.model.test.ts (hunk 11 above)
Comment (thehabbos007) [reply packages/hash/integration/src/tests/model/knowledge/link.model.test.ts]: Addressed in c3791a565d1b70a5f1813a42cc3bf1085ff25a47

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 7 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts:133]: It isn't guaranteed, and could be an array of links of the same `Friend` type. We don't support inserting multiple types of the same type yet, though (the whole 1:N)

**Review by TimDiekmann — COMMENTED** [review]
LGTM, but leaving it for @Alfred-Mountfield to resolve his comment chains and approve the PR 🙂

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 3 above)
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/knowledge/link.model.ts:51]: Apologies I had failed to notice it was static 🙃 

I don't have a strong opinion on your suggestion, if you think it'd be a good idea we can do so, or we can experiment with it at a later date

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 4 above)
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/knowledge/link.model.ts:106]: Gotcha, I don't really mind us calling it `1:N` but I have more issues with phrasings like: "the link has many targets" because it's not a single link.

To be clear I'm not just being pedantic about phrasing here, I was concerned it was affecting the design of this model class. Should a Model class for _a specific_ **link** be concerned with getting multiple links? If that's the pattern we have for `EntityModel` and such then I don't mind because it's consistent, but I was a bit concerned that it was related to the aforementioned point.

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 5 above)
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/knowledge/link.model.ts:108]: Gotcha, I didn't realise it was filtering out arrays. Wasn't sure if we had other stuff in here like `undefined` or some other stuff which felt like it should have been handled somewhere else haha

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 6 above)
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/knowledge/link.model.ts]: Thanks :)

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 7 above)
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/knowledge/link.model.ts:133]: Are we going to keep this method when we do support multiple links? Are we going to update this method and its docs (should we have a todo here then?)

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+     *   https://app.asana.com/0/1200211978612931/1202510174412974/f
+     */
+    const links = (
+      await LinkModel.getAllOutgoing(graphApi, {
+        sourceEntityModel,
+      })
+    ).filter(
+      (link) => link.linkTypeModel.schema.$id === linkTypeModel.schema.$id,
+    );
+    /** @todo this may return an array of links whe new support 1:N links. */
```
Comment (Alfred-Mountfield) [inline packages/hash/api/src/model/knowledge/link.model.ts:157]: Tiny nit, still a typo
```suggestion
    /** @todo this may return an array of links when we support 1:N links. */
```

Should we add an asana task here too

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 4 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts:106]: That's a valid concern. It is consistent with `EntityModel` fetching multiple entities (e.g. when calling `getAllLatest` `EntityModel[]` is returned, meaning multiple entities). 
I do think there will be some annoyances putting single and multi links in the same model class like we are doing here. Perhaps having a separate `MultiLinkModel` may make sense. As we don't need multi links right now, I don't think it's fair to make the assumption that it's useful to split them now, though.

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 3 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts:51]: I've added this task: https://app.asana.com/0/1202805690238892/1202890446280565/f to explore forcing proper usage of the model classes

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 7 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts:133]: Todo added 1f4b2c774ee8da8af2f7abffa7d7356491fd1a72

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 13 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts]: So sorry, I shouldn't have edited in GitHub :smile: 

713e9e57312790c52e5b86d6544b3252d36a9b87

**Review by Alfred-Mountfield — APPROVED** [review]

**Review by Alfred-Mountfield — APPROVED** [review]
Thanks for adding the links!
